src/shared/XLinkPrivateFields.c

```diff
@@ -65,13 +66,18 @@ streamId_t getStreamIdByName(xLinkDesc_t* link, const char* name)
 
 streamDesc_t* getStreamById(void* fd, streamId_t id)
 {
-    XLINK_RET_ERR_IF(id == INVALID_STREAM_ID, NULL);
     xLinkDesc_t* link = getLink(fd);
     XLINK_RET_ERR_IF(link == NULL, NULL);
     int stream;
     for (stream = 0; stream < XLINK_MAX_STREAMS; stream++) {
         if (link->availableStreams[stream].id == id) {
```

Comment:
Unsafe. `XLink_sem_wait()` must be before any field of `streamDesc_t` is read or write. There is no trick or lessening of the cost. On every read of any of the 32 entries of the array of `streamDesc_t`, the semaphore for that specific stream must be locked. This....is a mutex. This is one of the core problems with XLink...it misuses a semaphore and should be using a mutex.
I suspect you are trying to use an old method of optimizing called https://en.wikipedia.org/wiki/Double-checked_locking. That is an anti-pattern and is a source of problems.
One can't get around the need to lock a mutex. One has to pay the cost. The pattern everywhere in xlink should be...
```c++
// always check return result of XLink_sem_wait for os errors, destroyed sem by another thread, etc.
XLINK_RET_ERR_IF(XLink_sem_wait(...), NULL); // or some similar return value or if check

// do read, writes of fields in the `streamDesc_t` struct
...id,  ...name, ...packets, etc.

// same on post...also known as....unlock
XLINK_RET_ERR_IF(XLink_sem_post(&stream->sem), NULL);
```

Reply:
> This is one of the core problems with XLink...it misuses a semaphore and should be using a mutex.

I think that in this case the main issue is that we are calling this methods from another thread. I didn't go through whole XLink in this regard, but I think this should only be access by Scheduler thread otherwise.

> I suspect you are trying to use an old method of optimizing called https://en.wikipedia.org/wiki/Double-checked_locking. That is an anti-pattern and is a source of problems.

No, just not using the sem as a means of protection to the `stream` as if not checking the id before hand, its UB as it could be destroyed beforehand.
I think that with XLinkResetRemoteTimeout accessing this, we'd have to add a mutex for the access to stream contents instead. Or rather ditch `XLinkResetRemoteTimeout` in favor of a previous impl which doesn't access `closeLink` function in case of timeout, but tries unblocking the Scheduler instead (or event reader in that case)

Reply:
Yes, semaphores are very useful and good when signaling between threads. But that is not here.
Here, we have a single thread that needs exlusive access to the `streamDesc_t` struct to read+compare its `id` member field.
Here is the code I suggest, which can be similarly applied to `getStreamByName()`
```c++
streamDesc_t* getStreamById(void* fd, streamId_t id)
{
    XLINK_RET_ERR_IF(id == INVALID_STREAM_ID, NULL);
    xLinkDesc_t* link = getLink(fd);
    XLINK_RET_ERR_IF(link == NULL, NULL);
    int stream;
    for (stream = 0; stream < XLINK_MAX_STREAMS; stream++) {
        // get exclusive access to the struct
        // if the semaphore/mutex was destroyed by another thread, skip this iteration
        // fyi: current xlink code always has 0 or 1 owners of stream semaphores (and I believe never should be > 1)
        if (XLink_sem_wait(&link->availableStreams[stream].sem) < 0) {
            continue;
        }
        // you could check for INVALID_STREAM_ID here but it is cheaper to put one test
        // at the beginning rather than repeatedly testing it here
        if (link->availableStreams[stream].id == id) {
            return &link->availableStreams[stream];
        }
        XLink_sem_post(&link->availableStreams[stream].sem);
    }
    return NULL;
}
```

I believe this approach brings the protection needed for the specific scenario for which you are interested...and use the existing xlink protection mechanisms/funcs...without cascading more code work.

🤔 I request we we separate/isolate so to resolve this specific PR scenario without expanding work. Yes...rearchitecting xlink thread safety for connections+streams could include a mutex for streams. I would prefer not to discuss that here as I have a separate long issue of bugs I've found, code examples, etc. That is in the realm of rearchitecting and would cascade work to other places outside the scope of this PR.